### PR TITLE
Implement Packages for Mac OS X 

### LINKED ISSUE #144: Implement Mac Packages
**Branch**  `master`

**Target** `0.15.x`

**Proposal Description**
To use [Packages](http://s.sudre.free.fr/Software/Packages/about.html) project in our `KA-Lite Monitor` and able to create a configurable installer. As of now we are using the [create-dmg](https://github.com/mrpau/create-dmg) project which we can't configure. 

**Checklist**
- [x] Able to display the `README` and `License`
- [x] Able to create a configurable installer.
- [x] Modify `setup.sh`
- [x] Update `README`

The image below demonstrates the use of  [Packages](http://s.sudre.free.fr/Software/Packages/about.html).
screenshot 2015-08-26 13 01 47

### PR DESCRIPTION
**Pull request by arceduardvincent**
Hi @aronasorman .

This will enhance the user experience using [Packages](http://s.sudre.free.fr/Software/Packages/about.html) 

**Enhancement Checklist :**

- [x] Able to display the `README` and `License`
- [x] Able to create a configurable installer.
- [x] Modify `setup.sh`
- [x] Update `README`

**Steps to create the installer:**

 1. Open terminal
 2. Navigate to your `./installer/osx` folder.
 3. run `./setup.sh "https://github.com/learningequality/ka-lite/archive/0.14.x.zip"`.
 4. open the build output `./installers/osx/temp/output/KA-Lite.mpkg`
 5. Install the newly built installer.

See the newly built `Installer` demonstration below.
installer-demo1

This fixes #144.

### CODE REVIEW

File: osx/KA-Lite-Packages/KA-Lite-Monitor/KA-Lite.pkgproj [hunk 1]
```diff
@@ -0,0 +1,901 @@
...
+		<key>PROJECT_PRESENTATION</key>
+		<dict>
+			<key>BACKGROUND</key>
+			<dict>
+				<key>ALIGNMENT</key>
+				<integer>6</integer>
+				<key>BACKGROUND_PATH</key>
+				<dict>
+					<key>PATH</key>
+					<string>/Users/arceduardvincent/installers/osx/KA-Lite-Packages/ka-lite-logo.png</string>
```
Comment (cpauya) [inline osx/KA-Lite-Packages/KA-Lite-Monitor/KA-Lite.pkgproj]: Why is the path hard-coded as absolute @arceduardvincent?  I thought this is supposed to use a relative path?

File: osx/KA-Lite-Packages/KA-Lite-Monitor/KA-Lite.pkgproj [hunk 2]
```diff
@@ -0,0 +1,901 @@
...
+			<dict>
+				<key>LOCALIZATIONS</key>
+				<array>
+					<dict>
+						<key>LANGUAGE</key>
+						<string>English</string>
+						<key>VALUE</key>
+						<dict>
+							<key>PATH</key>
+							<string>/Users/arceduardvincent/installers/osx/KA-Lite-Packages/Introduction.rtf</string>
```
Comment (cpauya) [inline osx/KA-Lite-Packages/KA-Lite-Monitor/KA-Lite.pkgproj]: Same as above, make the path relative.

File: osx/KA-Lite-Packages/KA-Lite-Monitor/KA-Lite.pkgproj [hunk 3]
```diff
@@ -0,0 +1,901 @@
...
+			<dict>
+				<key>LOCALIZATIONS</key>
+				<array>
+					<dict>
+						<key>LANGUAGE</key>
+						<string>English</string>
+						<key>VALUE</key>
+						<dict>
+							<key>PATH</key>
+							<string>/Users/arceduardvincent/installers/osx/KA-Lite-Packages/ReadMe.rtf</string>
```
Comment (cpauya) [inline osx/KA-Lite-Packages/KA-Lite-Monitor/KA-Lite.pkgproj]: Same as above, make path as relative.

File: osx/setup.sh [hunk 4]
```diff
@@ -303,56 +308,18 @@ if ! [ -d "$KA_LITE_MONITOR_APP_PATH" ]; then
     exit 2
 fi
 
-# sign the .app file
-# unlock the keychain first so we can access the private key
-# security unlock-keychain -p $KEYCHAIN_PASSWORD
-# codesign -s "$SIGNER_IDENTITY_APPLICATION" --force "$KA_LITE_MONITOR_APP_PATH"
```
Comment (cpauya) [inline osx/setup.sh]: Why did you remove the commented `codesign` command?  Do we not need it anymore?  We use it as a placeholder for now until we need it.

File: osx/setup.sh [hunk 5]
```diff
@@ -303,56 +308,18 @@ if ! [ -d "$KA_LITE_MONITOR_APP_PATH" ]; then
...
 
-# clone the .dmg builder if non-existent
-if ! [ -d $DMG_BUILDER_PATH ]; then
-    git clone https://github.com/mrpau/create-dmg.git $DMG_BUILDER_PATH
-fi
-
-# Remove the .dmg if it exists.
-test -e "$DMG_PATH" && rm "$DMG_PATH"
-# Add the README.md to the package.
-cp "$KA_LITE_README_PATH" "$RELEASE_PATH"
```
Comment (cpauya) [inline osx/setup.sh]: Did we add this to the packages project already?

File: osx/setup.sh [hunk 6]
```diff
@@ -303,56 +308,18 @@ if ! [ -d "$KA_LITE_MONITOR_APP_PATH" ]; then
...
-# fi
-
-echo "Done!"
-if [ -e "$DMG_PATH" ]; then
-    # codesign the built DMG file
-    # unlock the keychain first so we can access the private key
-    # security unlock-keychain -p $KEYCHAIN_PASSWORD
-    codesign -s "$SIGNER_IDENTITY_APPLICATION" --force "$DMG_PATH"
-    echo "You can now test the built installer at '$DMG_PATH'."
+# check if the `Packages` is installed
```
Comment (cpauya) [inline osx/setup.sh]: Comment is misleading, what do you mean "installed"?  The switch that you used below is `-e`.

File: osx/setup.sh [hunk 7]
```diff
@@ -303,56 +308,18 @@ if ! [ -d "$KA_LITE_MONITOR_APP_PATH" ]; then
...
+# check if the `Packages` is installed 
+if [ -e "$PACKAGES_EXC" ]; then
+    $PACKAGES_EXC $PACKAGES_PROJECT
+    rm -fr $OUTPUT_PATH/*
+    mv -v $PACKAGES_BUILD_FOLDER/* $OUTPUT_PATH
 else
-    echo "Sorry, something went wrong trying to build the installer at '$DMG_PATH'."
-    exit 1
-fi
+   echo "Sorry, something went wrong trying to build the installer at '$OUTPUT_PATH'."
```
Comment (cpauya) [inline osx/setup.sh]: Message is vague, why not show "`Packages` is not installed" as message instead?

**Comment by cpauya** [comment]
@arceduardvincent 

This PR assumes that the `Packages` is already installed on the build server or the user's computer, which is not safe.

I think the setup script must check if `Packages` app and do either of the following (like what we did with the to-be-removed [create-dmg](https://github.com/mrpau/create-dmg) package in `setup.sh`):

1. quit with a non-zero return value if it is not installed, or;
2. install the `Packages` app and continue

Is it easy to include a process to check if it is installed and quit and if not, install it inside the setup script?

**Comment by cpauya** [comment]
Good screenshot btw!  :+1:

However, I saw "Licecap" among the payload on your screenshot above, you must remove that and modify the package script accordingly.

screenshot 2015-08-27 20 42 12

@aronasorman Please hold on merging this yet. :smile_cat:

File: osx/setup.sh [hunk 8]
```diff
@@ -82,6 +82,14 @@ KA_LITE="ka-lite"
 KA_LITE_ZIP="$WORKING_DIR/$KA_LITE.zip"
 KA_LITE_DIR="$WORKING_DIR/$KA_LITE"
 
+
+PACKAGES_EXC="/usr/local/bin/packagesbuild"
+PACKAGES_PROJECT="$SCRIPTPATH/KA-Lite-Packages/KA-Lite-Monitor/KA-Lite.pkgproj"
+PACKAGES_BUILD_FOLDER="$SCRIPTPATH/KA-Lite-Packages/KA-Lite-Monitor/build"
+PACKAGES_DMG="Packages.dmg"
+PACKAGES_DMG_URL="http://s.sudre.free.fr/Software/files/$PACKAGES_DMG"
```
Comment (cpauya) [inline osx/setup.sh]: Do we still need the Packages URL here if we have installed it on the build server directly?

**Comment by arceduardvincent** [comment]
Hi @cpauya I was able cater your request above and you may check the update.

**Comment by cpauya** [comment]
This fixes #144.

File: osx/setup.sh [hunk 9]
```diff
@@ -14,7 +14,7 @@
 # 9. Create the `<Xcode_Resources>/ka-lite/kalite/local_settings.py` based on `local_settings.default`.
 # 10. Copy the `ka-lite` and `pyrun` folders to the Xcode Resources folder.
 # 11. Build the Xcode project to produce the .app.
-# 12. Build the .dmg.
+# 12. Build the .mpkg.
```
Comment (cpauya) [inline osx/setup.sh]: Please clean-up the comments because there are steps here that are no longer done like:

* # 6. Make ka-lite repo production-ready (delete .KALITE_SOURCE_DIR, etc).
* # 9. Create the `<Xcode_Resources>/ka-lite/kalite/local_settings.py` based on `local_settings.default`.

The download of the `assessment.zip` is also not included there.

**Comment by cpauya** [comment]
Stale, let's create another PR targeting `develop` or `0.16.x` @arceduardvincent.
